paasta_tools/api/auth_decorator.py

```diff
@@ -0,0 +1,145 @@
+import functools
+
+from bravado.docstring_property import docstring_property
+from bravado.exception import HTTPBadRequest
+
+import paasta_tools.api
+from paasta_tools.utils import load_system_paasta_config
+
+# from bravado_decorators.decorate_client import decorate_client
+
+
+class AuthFutureDecorator:
+
+    def __init__(self, future, cluster_name):
+        self.future = future
+        self.attempts = 0
+        self.cluster_name = cluster_name
+
+    def result(self, *args, **kwargs):
+        """ Wraps a Future's result method with a timer. """
+        try:
+            result = self.future.result(*args, **kwargs)
+        except HTTPBadRequest as e:
+            # Yes nginx responds with a 400 if the certificate
+            # is expired. So the best we can do is watch for that
+            # return code and check the body of the response
+            if 'SSL' in e.response.text:
+                if self.attempts >= 3:
+                    raise
+                self.attempts += 1
+                system_paasta_config = load_system_paasta_config()
+                paasta_tools.api.client.renew_issue_cert(
+                    system_paasta_config=system_paasta_config,
+                    cluster=self.cluster_name,
+                )
+                result = self.result(*args, **kwargs)
```

Comment:
It seems that on the first recursion call you'll try to get `result` from the same `self.future` object, right?
If that's true then it isn't clear why there will be a second HTTP request/response with the new cert/key and not just the same exception raised as the result of the previous future execution result.

Reply:
Yup, we discussed in real life. I think this is just bad naming, it definitely fires another request with the new cert.